contracts/exchange-libs/contracts/src/LibTransactionDecoder.sol

```diff
@@ -0,0 +1,191 @@
+            functionSelector == EXECUTE_TRANSACTION_SELECTOR
+            // TODO: add new noThrow cancel functions when https://github.com/0xProject/ZEIPs/issues/35 is merged.
+        ) {
+            revert("UNIMPLEMENTED");
+        } else {
+            revert("UNKNOWN_FUNCTION_SELECTOR");
+        }
+
+        if (functionSelector == BATCH_CANCEL_ORDERS_SELECTOR) {
+            // solhint-disable-next-line indent
+            orders = abi.decode(signedTransactionData.slice(4, signedTransactionData.length), (LibOrder.Order[]));
+            takerAssetFillAmounts = new uint256[](0);
+            signatures = new bytes[](0);
+        } else if (
+            functionSelector == BATCH_FILL_OR_KILL_ORDERS_SELECTOR ||
+            functionSelector == BATCH_FILL_ORDERS_NO_THROW_SELECTOR ||
+            functionSelector == BATCH_FILL_ORDERS_SELECTOR
+        ) {
+            (orders, takerAssetFillAmounts, signatures) = _makeReturnValuesForBatchFill(signedTransactionData);
+        } else if (functionSelector == CANCEL_ORDER_SELECTOR) {
+            orders = new LibOrder.Order[](1);
+            orders[0] = abi.decode(signedTransactionData.slice(4, signedTransactionData.length), (LibOrder.Order));
+            takerAssetFillAmounts = new uint256[](0);
+            signatures = new bytes[](0);
+        } else if (
+            functionSelector == FILL_OR_KILL_ORDER_SELECTOR ||
+            functionSelector == FILL_ORDER_SELECTOR ||
+            functionSelector == FILL_ORDER_NO_THROW_SELECTOR
+        ) {
+            (orders, takerAssetFillAmounts, signatures) = _makeReturnValuesForSingleOrderFill(signedTransactionData);
+        } else if (
+            functionSelector == MARKET_BUY_ORDERS_SELECTOR ||
+            functionSelector == MARKET_BUY_ORDERS_NO_THROW_SELECTOR ||
+            functionSelector == MARKET_SELL_ORDERS_SELECTOR ||
+            functionSelector == MARKET_SELL_ORDERS_NO_THROW_SELECTOR
+        ) {
+            (orders, takerAssetFillAmounts, signatures) = _makeReturnValuesForMarketFill(signedTransactionData);
+        } else if (functionSelector == MATCH_ORDERS_SELECTOR) {
+            (
+                LibOrder.Order memory leftOrder,
+                LibOrder.Order memory rightOrder,
+                bytes memory leftSignature,
+                bytes memory rightSignature
+            ) = abi.decode(
+                signedTransactionData.slice(4, signedTransactionData.length),
+                (LibOrder.Order, LibOrder.Order, bytes, bytes)
+            );
+
+            orders = new LibOrder.Order[](2);
+            orders[0] = leftOrder;
+            orders[1] = rightOrder;
+
+            takerAssetFillAmounts = new uint256[](2);
+            takerAssetFillAmounts[0] = leftOrder.takerAssetAmount;
+            takerAssetFillAmounts[1] = rightOrder.takerAssetAmount;
+
+            signatures = new bytes[](2);
+            signatures[0] = leftSignature;
+            signatures[1] = rightSignature;
+        }
```

Comment:
Nit - could be good to add an `else` here as well that reverts. There are so many selectors now; this way we'll know if one is ever unhandled.

Reply:
I declined to catch unknown selectors in this `if` because I already set up such an else in the previous `if` in this function.